core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -618,10 +618,42 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
   /**
    * Provides some of the environment required to run this effect,
    * leaving the remainder `R0`.
+   *
+   * {{{
+   * val effect: ZManaged[Console with Logging, Nothing, Unit] = ???
+   *
+   * effect.provideSome[Console](env =>
+   *   new Console with Logging {
+   *     val console = env.console
+   *     val logging = new Logging.Service[Any] {
+   *       def log(line: String) = console.putStrLn(line)
+   *     }
+   *   }
+   * )
+   * }}}
    */
   def provideSome[R0](f: R0 => R)(implicit ev: NeedsEnv[R]): ZManaged[R0, E, A] =
     ZManaged(reserve.provideSome(f).map(r => Reservation(r.acquire.provideSome(f), e => r.release(e).provideSome(f))))
 
+  /**
+   * An effectful version of `provideSome`, useful when the act of partial
+   * provision requires an effect.
+   *
+   * {{{
+   * val effect: ZManaged[Console with Logging, Nothing, Unit] = ???
+   *
+   * val r0: URIO[Console, Console with Logging] = ???
+   *
+   * effect.provideSomeM(r0)
+   * }}}
+   */
+  def provideSomeM[R0, E1 >: E](
+    f: ZIO[R0, E1, R]
```

Comment:
I think we should say that the effect cannot fail here.

Reply:
I thought about that. I thought about going that route. I was wondering if it was too restrictive. But then again, accepting an effect that can fail, with a die hidden under the hood for the release, might lead to even yet more unexpected results, plus, it is binding to a subtype of throwable.

Reply:
Actually there is some weirdness here because we are running the effect twice. Why don't we implement this as `f.toManaged_.flatMap(self.provide)`? Then we can also remove the constraint on the error type.

Reply:
Looks great now! Very minor suggestion, maybe call the parameter `r0` instead of `f` for consistency with same method on ZIO?